fdbserver/storageserver.actor.cpp

```diff
@@ -3644,25 +3660,67 @@ ACTOR Future<Void> checkBehind( StorageServer* self ) {
 	}
 }
 
+ACTOR Future<Void> serveGetValueRequests( StorageServer* self, FutureStream<GetValueRequest> getValue ) {
+	loop {
+		GetValueRequest req = waitNext(getValue);
+		// Warning: This code is executed at extremely high priority (TaskPriority::LoadBalancedEndpoint), so downgrade before doing real work
+		if( req.debugID.present() )
+			g_traceBatch.addEvent("GetValueDebug", req.debugID.get().first(), "storageServer.received"); //.detail("TaskID", g_network->getCurrentTask());
+
+		if (SHORT_CIRCUT_ACTUAL_STORAGE && normalKeys.contains(req.key))
+			req.reply.send(GetValueReply());
+		else
+			self->actors.add(self->readGuard(req , getValueQ));
+	}
+}
+
+ACTOR Future<Void> serveGetKeyValuesRequests( StorageServer* self, FutureStream<GetKeyValuesRequest> getKeyValues ) {
+	loop {
+		GetKeyValuesRequest req = waitNext(getKeyValues);
+		// Warning: This code is executed at extremely high priority (TaskPriority::LoadBalancedEndpoint), so downgrade before doing real work
+		self->actors.add(self->readGuard(req, getKeyValuesQ));
+	}
+}
+
+ACTOR Future<Void> serveGetKeyRequests( StorageServer* self, FutureStream<GetKeyRequest> getKey ) {
+	loop {
+		GetKeyRequest req = waitNext(getKey);
+		// Warning: This code is executed at extremely high priority (TaskPriority::LoadBalancedEndpoint), so downgrade before doing real work
+		self->actors.add(self->readGuard(req , getKeyQ));
+	}
+}
+
+ACTOR Future<Void> serveWatchValueRequests( StorageServer* self, FutureStream<WatchValueRequest> watchValue ) {
+	loop {
+		WatchValueRequest req = waitNext(watchValue);
+		// TODO: fast load balancing?
+		// SOMEDAY: combine watches for the same key/value into a single watch
+		self->actors.add(self->readGuard(req, watchValueQ));
+	}
+}
+
 ACTOR Future<Void> storageServerCore( StorageServer* self, StorageServerInterface ssi )
 {
 	state Future<Void> doUpdate = Void();
 	state bool updateReceived = false;  // true iff the current update() actor assigned to doUpdate has already received an update from the tlog
-	state ActorCollection actors(false);
 	state double lastLoopTopTime = now();
 	state Future<Void> dbInfoChange = Void();
 	state Future<Void> checkLastUpdate = Void();
 	state Future<Void> updateProcessStatsTimer = delay(SERVER_KNOBS->FASTRESTORE_UPDATE_PROCESS_STATS_INTERVAL);
 
-	actors.add(updateStorage(self));
-	actors.add(waitFailureServer(ssi.waitFailure.getFuture()));
-	actors.add(self->otherError.getFuture());
-	actors.add(metricsCore(self, ssi));
-	actors.add(logLongByteSampleRecovery(self->byteSampleRecovery));
-	actors.add(checkBehind(self));
+	self->actors.add(updateStorage(self));
```

Comment:
Why separating these `when` statements into individual actors can improve the performance?

Reply:
it will probably have a pretty minor effect, but every time a message comes in on any of the promise streams we have to evaluate all of the when block again the next time around the loop